### ISSUE TITLE
Discover returns an error

### ISSUE THREAD

**Issue by victorjaviermartin** [mobile-dev-inc/dadb#37]
Hi guys,

I'm Victor, and I'm trying to run the library in a poc application to know how to use it. But the first method that I'm trying is "discover", but it returning me a FileNotFoundException

I'm doing:

```
fun discover(){
        try {
            val result = Dadb.discover("localhost")
            binding.tvcommandline.text = "" + result.toString()
        }catch (ex: java.lang.RuntimeException){
            binding.tvcommandline.text = "Failed to discover emulator. No adb device found"
            ex.printStackTrace()
        }
    }

```
The error is:

`Caused by: java.io.FileNotFoundException: .android/adbkey: open failed: ENOENT (No such file or directory)`

Some advice?

I'm testing on a Samsung S20 Ultra with Android 12 and connected via USB.

Thank you

**Comment by sharpordie** [comment]
Dadb doesn't create the keypair automatically on Android device. You have to create those keys manually and use them the discover method.

This is a code sample to create them in your app's temp directory.

```kotlin
class YourViewModel(application: Application) : AndroidViewModel(application) {

    val context by lazy { getApplication<Application>().applicationContext }

    suspend fun keygen(refresh: Boolean = false): AdbKeyPair = withContext(IO) {
        val tempdir = context.cacheDir
        val pvtFile = File(tempdir, "adbkey")
        val pubFile = File(tempdir, "adbkey.pub")
        if (refresh || !pubFile.exists() || !pvtFile.exists()) {
            if (pubFile.exists()) pubFile.delete()
            if (pvtFile.exists()) pvtFile.delete()
            AdbKeyPair.generate(pvtFile, pubFile)
        }
        AdbKeyPair.read(pvtFile, pubFile)
    }
    
    suspend fun onButtonClicked() = viewModelScope.launch {
        val handler = Dadb.discover("localhost", keygen())
    }        
}        
```

**Comment by victorjaviermartin** [comment]
> 
Hi @sharpordie with those lines, discover returns me a "null"

**Comment by Leland-Takamine** [comment]
@victorjaviermartin Can you please paste the whole stack trace?

@sharpordie Dadb does actually generate the adb key automatically now: https://github.com/mobile-dev-inc/dadb/blob/42a0125c47b196418fffdf4cdf6fdc04935e523e/dadb/src/main/kotlin/dadb/AdbKeyPair.kt#L77

**Comment by victorjaviermartin** [comment]
Yes @sharpordie. I need to test again to take the full stacktrace, but I don't have time at the moment. Sorry.

**Comment by sharpordie** [comment]
Sorry for the long waiting time.

Here is a working project using an old version (0.0.11) of Dadb :
https://github.com/sharpordie/hisendal

I quickly tested with version 1.2.4, but it doesn't work unfortunately.
I hope this helps you anyway.

**Comment by sharpordie** [comment]
@victorjaviermartin
Well, this should work fine with the latest version of Dadb.
https://github.com/sharpordie/hisendal

It's still using the ugly GlobalScope.async. This is used to throw an exception if the target device has not accepted the connection yet and prevents it [from blocking endlessly](https://github.com/mobile-dev-inc/dadb/issues/21).


I'm not working with Dadb anymore for the moment, my initial goal was to use Dadb on Android and iOS through Kotlin Multiplatform and use it in my Flutter projects with Klutter. But that's way too much for my narrow brain. The fact that Dadb doesn't allow to pair with Android 11+ devices is also a serious problem.

**Comment by victorjaviermartin** [comment]
Thank you @sharpordie but unfortunately this is obsolete from android 11+. Thank you for your effort.
